lib/integrations/awsoidc/idp_iam_config.go

```diff
@@ -104,18 +176,102 @@ type IdPIAMConfigureClient interface {
 }
 
 type defaultIdPIAMConfigureClient struct {
+	httpClient *http.Client
+
 	*iam.Client
+	awsConfig aws.Config
 	stsClient *sts.Client
+	s3Client  *s3.Client
 }
 
 // GetCallerIdentity returns details about the IAM user or role whose credentials are used to call the operation.
-func (d defaultIdPIAMConfigureClient) GetCallerIdentity(ctx context.Context, params *sts.GetCallerIdentityInput, optFns ...func(*sts.Options)) (*sts.GetCallerIdentityOutput, error) {
+func (d *defaultIdPIAMConfigureClient) GetCallerIdentity(ctx context.Context, params *sts.GetCallerIdentityInput, optFns ...func(*sts.Options)) (*sts.GetCallerIdentityOutput, error) {
 	return d.stsClient.GetCallerIdentity(ctx, params, optFns...)
 }
 
+// CreateBucket creates an Amazon S3 bucket.
+func (d *defaultIdPIAMConfigureClient) CreateBucket(ctx context.Context, params *s3.CreateBucketInput, optFns ...func(*s3.Options)) (*s3.CreateBucketOutput, error) {
+	return d.s3Client.CreateBucket(ctx, params, optFns...)
+}
+
+// PutObject adds an object to a bucket.
+func (d *defaultIdPIAMConfigureClient) PutObject(ctx context.Context, params *s3.PutObjectInput, optFns ...func(*s3.Options)) (*s3.PutObjectOutput, error) {
+	return d.s3Client.PutObject(ctx, params, optFns...)
+}
+
+// HeadBucket adds an object to a bucket.
+func (d *defaultIdPIAMConfigureClient) HeadBucket(ctx context.Context, params *s3.HeadBucketInput, optFns ...func(*s3.Options)) (*s3.HeadBucketOutput, error) {
+	return d.s3Client.HeadBucket(ctx, params, optFns...)
+}
+
+// PutBucketPolicy applies an Amazon S3 bucket policy to an Amazon S3 bucket.
+func (d *defaultIdPIAMConfigureClient) PutBucketPolicy(ctx context.Context, params *s3.PutBucketPolicyInput, optFns ...func(*s3.Options)) (*s3.PutBucketPolicyOutput, error) {
+	return d.s3Client.PutBucketPolicy(ctx, params, optFns...)
+}
+
+// DeletePublicAccessBlock removes the PublicAccessBlock configuration for an Amazon S3 bucket.
+func (d *defaultIdPIAMConfigureClient) DeletePublicAccessBlock(ctx context.Context, params *s3.DeletePublicAccessBlockInput, optFns ...func(*s3.Options)) (*s3.DeletePublicAccessBlockOutput, error) {
+	return d.s3Client.DeletePublicAccessBlock(ctx, params, optFns...)
+}
+
+// GetBucketPolicy returns the policy of a specified bucket
+func (d *defaultIdPIAMConfigureClient) GetBucketPolicy(ctx context.Context, params *s3.GetBucketPolicyInput, optFns ...func(*s3.Options)) (*s3.GetBucketPolicyOutput, error) {
+	return d.s3Client.GetBucketPolicy(ctx, params, optFns...)
+}
+
+// CreateBucketRegion returns the region where the bucket should be created.
+func (d *defaultIdPIAMConfigureClient) CreateBucketRegion() string {
+	return d.awsConfig.Region
+}
+
+// SetAWSRegion sets the aws region for next api calls.
+func (d *defaultIdPIAMConfigureClient) SetAWSRegion(awsRegion string) {
+	d.awsConfig.Region = awsRegion
+
+	// S3 Client is the only client that depends on the region.
+	d.s3Client = s3.NewFromConfig(d.awsConfig)
```

Comment:
As I understand client is basically supposed to be used as a one-off thing, within one goroutine, so we probably don't need any locking around it. But maybe we can add comment to the default client that it's not thread-safe?